File: leetcode-04/solutions/cpp/0117.cpp
class Solution {
 public:
  Node* connect(Node* root) {
    Node* node = root;
    Node* curr = new Node(NULL);
    Node* prev = curr;

    while (node) {
      while (node) {
        curr->next = node->left;
        if (curr->next) curr = curr->next;
        curr->next = node->right;
        if (curr->next) curr = curr->next;
        node = node->next;
      }
      node = prev->next;
      curr = prev;
    }

    return root;
  }
};

File: leetcode-04/solutions/python3/0117.py
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        node = root
        curr = Node(None, None, None, None)
        prev = curr

        while node:
            while node:
                curr.next = node.left
                curr = curr.next or curr
                curr.next = node.right
                curr = curr.next or curr
                node = node.next
            node = prev.next
            curr = prev

        return root
